org.eclipse.jdt.core.compiler.batch/src/org/eclipse/jdt/internal/compiler/lookup/InferenceContext18.java

```diff
@@ -1910,4 +1911,12 @@ public void cleanUp() {
 		this.b2 = null;
 		this.currentBounds = null;
 	}
+
+	public boolean isInexactVarargsInference() {
+		return this.isInexactVarargsInference;
+	}
+
+	public void setInexactVarargsInference(boolean isInexactVarargsInference) {
```

Comment:
> Or will the affected InferenceContext18 be discarded and replaced with a fresh one at some point?

Yes, the inexact IC18 will be discarded, and eventually replaced by a new round of IC18.

https://github.com/eclipse-jdt/eclipse.jdt.core/blob/4b85299549560d80141cd654cd280086577cf3f1/org.eclipse.jdt.core.compiler.batch/src/org/eclipse/jdt/internal/compiler/ast/MessageSend.java#L1008-L1029

Usually the inexact varargs inference occurs in the first round of inference in **L1018**, where the method arguments could be improper types, leading inexact varargs inference result.

It then triggers a second round of resolving with improved target types from the methods parameters, see **L1027** with `resolvePolyExpressionArguments`.

Reply:
> Once set to `true` this flag is never reset. I wonder if we should? Would that enable us to reason that anything that has initially been skipped has been evaluated eventually?
> 
> Or will the affected InferenceContext18 be discarded and replaced with a fresh one at some point?

To answer this question, I'm thinking under what situation this exact varargs method inference would happen.

Here is a typical use case for it (e.g. `messageSend(arg -> varargsPGMB(arg))`).
- In the first round of resolving method binding for [messageSend#L1018](https://github.com/eclipse-jdt/eclipse.jdt.core/blob/master/org.eclipse.jdt.core.compiler.batch/src/org/eclipse/jdt/internal/compiler/ast/MessageSend.java#L1018), since one argument is a poly lambda expression, it uses the original InferenceVariable (not proper type) from lambda definition as InferenceSubstitution to infer lambda and its nested expressions. This is how the inner varargs expression gets inexact varargs inference..
- Then `messageSend` triggers a second round of resolving with improved target types from the methods parameters, see `resolvePolyExpressionArguments` in [MessageSend#L1027](https://github.com/eclipse-jdt/eclipse.jdt.core/blob/master/org.eclipse.jdt.core.compiler.batch/src/org/eclipse/jdt/internal/compiler/ast/MessageSend.java#L1027). This time the InferenceVariable argument would be replaced by TypeVariableBinding (proper type), and can give a more accurate inference for the inner varargs method.

@stephan-herrmann  WDYT? Any other examples do you think it would fail?

Reply:
@testforstephen  sorry, now I had lost this pr from the radar.

I didn't find an example that contradicts your explanation, but I don't think the local reasoning about `MessageSend.findMethodBinding()` will grant us real certainty, given that even `ASTNode.resolvePolyExpressionArguments()` (which is designed to wrap up details of this message send) can be invoked with `isInexactVarargsInference` set, as witnessed by your very change in that method. After `resolvePolyExpressionArguments()` has run, who asserts that we will see yet another round of inference?

Zooming out I suggest to correlate `isInexactVarargsInference` against `IC18.stepCompleted == TYPE_INFERRED_FINAL`. The following assertion would raise an alarm if ever we consider a round of inference as complete and final despite that flag being set:
```
diff --git a/org.eclipse.jdt.core.compiler.batch/src/org/eclipse/jdt/internal/compiler/lookup/InferenceContext18.java b/org.eclipse.jdt.core.compiler.batch/src/org/eclipse/jdt/internal/compiler/lookup/InferenceContext18.java
index 0171e2f..91b86c7 100644
--- a/org.eclipse.jdt.core.compiler.batch/src/org/eclipse/jdt/internal/compiler/lookup/InferenceContext18.java
+++ b/org.eclipse.jdt.core.compiler.batch/src/org/eclipse/jdt/internal/compiler/lookup/InferenceContext18.java
@@ -483,6 +483,7 @@
                                this.currentBounds = solution; // this is final, keep the result:
                        return solution;
                } finally {
+                       assert !(step == TYPE_INFERRED_FINAL && this.isInexactVarargsInference);
                        this.stepCompleted = step;
                }
        }
```

Looking at usage of `TYPE_INFERRED_FINAL` this looks like a good candidate to distinguish preliminary vs. final inference.

WDYT?

Reply:
PR updated. Adding an `assert` seems to be a good approach to ensure the inexactVarargsInference is not set in the final round of inference.

One question is that Eclipse does not enable `-ea` by default, so how can we guarantee that the assert code can take effect in real production environments?

Reply:
Once again I dropped the ball. Hint: by requesting a review you may ensure that a PR remains on my radar :)

> One question is that Eclipse does not enable -ea by default, so how can we guarantee that the assert code can take effect in real production environments?

`-ea` is indeed passed for tests, but regarding production environments I don't think throwing `AssertionError` would be nice communication with the user :) . For that reason I'd be fine with enabling it only on request, in particular when a bug is reported that could potentially relate to this issue.

OK, now would be a good point in time to wrap up here. @testforstephen, if you once more resolve merge conflicts, I promise to merge it without requesting further back and forth :)